gecode/int/branch/cbs.hpp

```diff
@@ -0,0 +1,244 @@
+ *
+ *  THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+ *  EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+ *  MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+ *  NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE
+ *  LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
+ *  OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION
+ *  WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+ *
+ */
+
+#ifdef GECODE_HAS_CBS
+
+namespace Gecode { namespace Int { namespace Branch {
+
+  template<class View>
+  forceinline void
+  CBSBrancher<View>::VarIdToPos::init() {
+    assert(object() == nullptr);
+    object(new VarIdToPosO());
+  }
+
+  template<class View>
+  forceinline bool
+  CBSBrancher<View>::VarIdToPos::isIn(unsigned int var_id) const {
+    auto *hm = &static_cast<VarIdToPosO*>(object())->_varIdToPos;
+    return hm->find(var_id) != hm->end();
+  }
+
+  template<class View>
+  forceinline int
+  CBSBrancher<View>::VarIdToPos::operator[](unsigned int i) const {
+    return static_cast<VarIdToPosO*>(object())->_varIdToPos.at(i);
+  }
+
+  template<class View>
+  forceinline void
+  CBSBrancher<View>::VarIdToPos::insert(unsigned int var_id,
+                                             unsigned int pos) {
+    static_cast<VarIdToPosO*>(object())
+      ->_varIdToPos.insert(std::make_pair(var_id, pos));
+  }
+
+  template<class View>
+  CBSBrancher<View>::CBSBrancher(Home home, ViewArray<View>& x0)
+    : Brancher(home), x(x0),
+      logProp(typename decltype(logProp)::size_type(),
+              typename decltype(logProp)::hasher(),
+              typename decltype(logProp)::key_equal(),
+              typename decltype(logProp)::allocator_type(home)) {
+    varIdToPos.init();
+    for (int i=0; i<x.size(); i++)
+      varIdToPos.insert(x[i].id(), i);
+  }
+
+  template<class View>
+  forceinline void
+  CBSBrancher<View>::post(Home home, ViewArray<View>& x) {
+    (void) new (home) CBSBrancher(home,x);
+  }
+
+  template<class View>
+  Actor* CBSBrancher<View>::copy(Space& home) {
+    return new (home) CBSBrancher(home,*this);
+  }
+
+  template<class View>
+  forceinline size_t
+  CBSBrancher<View>::dispose(Space& home) {
+    home.ignore(*this, AP_DISPOSE);
+    (void) Brancher::dispose(home);
+    return sizeof(*this);
+  }
+
+  template<class View>
+  CBSBrancher<View>::CBSBrancher(Space& home, CBSBrancher& b)
+    : Brancher(home,b),
+      varIdToPos(b.varIdToPos),
+      logProp(b.logProp.begin(), b.logProp.end(),
+              typename decltype(logProp)::size_type(),
+              typename decltype(logProp)::hasher(),
+              typename decltype(logProp)::key_equal(),
+              typename decltype(logProp)::allocator_type(home)) {
+    x.update(home,b.x);
+  }
+
+  template<class View>
+  bool CBSBrancher<View>::status(const Space& home) const {
+    auto& h = const_cast<Space&>(home);
```

Comment:
There is no method for iterating in const propagators, so I needed to do this. In line 120, I tried to make explicit the fact that domainsizesum is a const method.

Reply:
I just changed that in the kernel itself...

Reply:
Which means you can remove the cast eventually